java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java

```diff
@@ -0,0 +1,288 @@
+    }
+
+    for (String name : left.getChildFieldNames()) {
+      RangeEqualsVisitor visitor = new RangeEqualsVisitor(rightVector.getChild(name),
+          leftStart, rightStart, length);
+      if (!left.getChild(name).accept(visitor)) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  protected boolean compareBaseFixedWidthVectors(BaseFixedWidthVector left) {
+
+    if (!compareValueVector(left, right)) {
+      return false;
+    }
+
+    for (int i = 0; i < length; i++) {
+      int leftIndex = leftStart + i;
+      int rightIndex = rightStart + i;
+
+      boolean isNull = left.isNull(leftIndex);
+
+      if (isNull != right.isNull(rightIndex)) {
+        return false;
+      }
+
+      int typeWidth = left.getTypeWidth();
+      if (!isNull) {
+        int startByteLeft = typeWidth * leftIndex;
+        int endByteLeft = typeWidth * (leftIndex + 1);
+
+        int startByteRight = typeWidth * rightIndex;
+        int endByteRight = typeWidth * (rightIndex + 1);
+
+        int ret = ByteFunctionHelpers.equal(left.getDataBuffer(), startByteLeft, endByteLeft,
+            right.getDataBuffer(), startByteRight, endByteRight);
+
+        if (ret == 0) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  protected boolean compareBaseVariableWidthVectors(BaseVariableWidthVector left) {
+    if (!compareValueVector(left, right)) {
+      return false;
+    }
+
+    for (int i = 0; i < length; i++) {
+      int leftIndex = leftStart + i;
+      int rightIndex = rightStart + i;
+
+      boolean isNull = left.isNull(leftIndex);
+      if (isNull != right.isNull(rightIndex)) {
+        return false;
+      }
+
+      int offsetWidth = BaseVariableWidthVector.OFFSET_WIDTH;
+
+      if (!isNull) {
+        final int startByteLeft = left.getOffsetBuffer().getInt(leftIndex * offsetWidth);
+        final int endByteLeft = left.getOffsetBuffer().getInt((leftIndex + 1) * offsetWidth);
+
+        final int startByteRight = right.getOffsetBuffer().getInt(rightIndex * offsetWidth);
+        final int endByteRight = right.getOffsetBuffer().getInt((rightIndex + 1) * offsetWidth);
+
+        int ret = ByteFunctionHelpers.equal(left.getDataBuffer(), startByteLeft, endByteLeft,
+            right.getDataBuffer(), startByteRight, endByteRight);
+
+        if (ret == 0) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  protected boolean compareListVectors(ListVector left) {
+    if (!compareValueVector(left, right)) {
+      return false;
+    }
+
+    for (int i = 0; i < length; i++) {
+      int leftIndex = leftStart + i;
+      int rightIndex = rightStart + i;
+
+      boolean isNull = left.isNull(leftIndex);
+      if (isNull != right.isNull(rightIndex)) {
+        return false;
+      }
+
+      int offsetWidth = BaseRepeatedValueVector.OFFSET_WIDTH;
+
+      if (!isNull) {
+        final int startByteLeft = left.getOffsetBuffer().getInt(leftIndex * offsetWidth);
```

Comment:
nit (pls ignore if you don't agree): names are a bit confusing - startIndexLeft/endIndexLeft ?

Reply:
Thanks, would also be resolved together in follow-ups.